### DISCUSSION TITLE
[Chapter 4] Q & A

### DISCUSSION THREAD

**Discussion by JoisFe** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#41] [CHAPTER 4 스트림 소개 - Q&A] [answered]
# [Chapter 4] Q & A

### Q3) 스트림의 중간 연산 (intermediate operation)은 지연 연산(lazy evaluation) 의 특징을 가지는데 왜 이러한 특징을 가지게 하며, 이로 인한 장점은? (또한 파이썬의 Iterator와 비슷한 특징?)

**Comment by ahngilwoong** [comment, marked as answer]
Q1) 빌딩 블록이란 하나의 컴퓨터 시스템이나 대형 소프트웨어를 설계하는 과정에서 이들 제품을 구성하는 각각의 구성 요소를 서로 독립된 모듈로써 구성하는 방법이라고 합니다. 여기서 고수준이라는 말은 어려운 수준이란 뜻보단 추상화의 정도를 나타내는 말로 개인적인 의견으로는 두개의 뜻을 합친 풀이로 해석할 수 있을 것 같습니다. 특정 스레딩 모델에 제한되지 않는 정확한 이유는 추측만 할뿐 명확하지 않아 저도 다른 분들의 다양한 의견을 듣고 싶습니다.

- 고수준에 대한 내용
https://meetup.nhncloud.com/posts/345

- 빌딩 블록 출처
https://wordrow.kr/%EC%9E%90%EC%84%B8%ED%95%9C-%EC%9D%98%EB%AF%B8/%EB%B9%8C%EB%94%A9%20%EB%B8%94%EB%A1%9D%20%EB%B0%A9%EC%8B%9D/3/


Q2) 말씀해주신 라이브러리들은 사용한적이 없습니다.. 
하지만 그 뒤 질문이 공교롭게도 얼마 전 회사 선배님이 외부 라이브러리를 선택하는 기준에 대해 말씀해주신적이 있어 그 내용을 써보도록 하겠습니다. 
(회사 내에서는 어떠한 라이브러리도 사실상 제 마음대로 못 쓰지만 기준에 대해서만 말씀해주셨습니다..)

1. 유지보수와 업데이트 (라이브러리들이 믿을만한건지 버그 픽스, 보안 이슈등을 수시로 업데이트하며 유지보수가 되고있는지?)
2. 라이브러리에 대한 doc 설명 (가져다 쓰기 쉽게 잘 설명되어있는지)
3. 성능과 안정성 (실무에서 안정적으로 사용되어 이미 다른 곳도 많이 쓰고 충분한 검증이 되었는지)

크게 이 3가지에 대한 내용이 인상적이었습니다.

Q3) 지연 연산 자체의 장점이 그 특징의 이유인 것 같습니다. 지연 연산은 필요한 시점에서만 연산을 수행하기 때문에 불필요한 연산을 피할 수 있는데 연산을 미뤄놓고 최종 연산시에만 필요한 데이터만 처리하여 낭비를 줄인 듯 합니다!

**Comment by Irisation23** [comment]
Q3) 스트림의 중간 연산 (intermediate operation)은 지연 연산(lazy evaluation) 의 특징을 가지는데 왜 이러한 특징을 가지게 하며, 이로 인한 장점은? (또한 파이썬의 Iterator와 비슷한 특징?)

### 중간 연산
- 연산 결과가 Stream인 연산 
- n번 사용 가능

### 지연 연산
- 최종 연산을 마쳐야 중간 연산을 실행함.

`@ahngilwoong`님과 의견이 같습니다.
